File: Java/jstructs/List.java
package jstructs;

/**
 * <p>
 * === List ===
 * </p><p>
 * Date : May 02, 2017
 * </p><p>
 * TODO Description here
 * </p>
 *
 * @author Michael van Dyk
 */
public interface List<E> extends Deque<E> {

    int size();
    E get(int index);
    boolean set(int index, E item);
    boolean add(int index, E item);
    E remove(int index);

    @Override
    default boolean addFirst(E item) {
        return (add(0, item));
    }

    @Override
    default E removeFirst() {
        return (remove(0));
    }

    @Override
    default boolean addLast(E item) {
        return (add(size(), item));
    }

    @Override
    default E removeLast() {
        return (remove(size() - 1));
    }
}

File: Kotlin/kstructs/list/List.kt
package kstructs.list

import kstructs.Deque

/**
 * <p>
 * === List ===
 * </p><p>
 * Date : May 02, 2017
 * </p><p>
 * TODO Description here
 * </p>
 * @author Michael van Dyk
 */
interface List<E> : Deque<E> {
    operator fun get(index: Int): E?
    operator fun set(index: Int, item: E): E?
    fun add(index: Int, item: E): Boolean
    fun remove(index: Int): E?

    override fun addFirst(item: E) = add(0, item)
    override fun removeFirst() = remove(0)
    override fun addLast(item: E) = add(size(), item)
    override fun removeLast() = remove(size() - 1)
}